yt_dlp/extractor/tubetugraz.py

```diff
@@ -0,0 +1,349 @@
+
+    def _guess_formats(self, formats, format_types, id):
+        PREFERRED_TYPE = "presentation"
+
+        for type in ("presentation", "presenter"):
+            m3u8_url = "https://wowza.tugraz.at/matterhorn_engage/smil:engage-player_%s_%s.smil/playlist.m3u8" % (id, type)
+            mpd_url = "https://wowza.tugraz.at/matterhorn_engage/smil:engage-player_%s_%s.smil/manifest_mpm4sav_mvlist.mpd" % (id, type)
+
+            if type == PREFERRED_TYPE:
+                preference = -1
+            else:
+                preference = -2
+
+            if "hls" not in format_types[type]:
+                hls_page = self._download_webpage_handle(
+                    m3u8_url, None,
+                    note="guessing location of %s HLS manifest" % type,
+                    errnote=False,
+                    fatal=False)
+                if hls_page is not False:
+                    hls_formats = self._extract_m3u8_formats(
+                        m3u8_url, None,
+                        note="downloading %s HLS manifest" % type,
+                        fatal=False,
+                        ext="mp4")
+
+                    for format in hls_formats:
+                        format["preference"] = preference
+                        self._gen_format_id(format, type, "hls", format_types)
+
+                    formats.extend(hls_formats)
+            if "dash" not in format_types[type]:
+                dash_page = self._download_webpage_handle(
+                    mpd_url, None,
+                    note="guessing location of %s DASH manifest" % type,
+                    errnote=False,
+                    fatal=False)
+                if dash_page is not False:
+                    dash_formats = self._extract_mpd_formats(
+                        mpd_url, None,
+                        note="downloading %s DASH manifest" % type,
+                        fatal=False)
+
+                    for format in dash_formats:
+                        format["preference"] = preference
+                        self._gen_format_id(format, type, "dash", format_types)
+
+                    formats.extend(dash_formats)
+
+    def _gen_format_id(self, format, type, transport, format_types):
+        format_types[type][transport] += 1
+        index = format_types[type][transport]
+
+        vbr = format.get("vbr")
+        abr = format.get("abr")
+        tbr = format.get("tbr") or ((vbr or 0) + (abr or 0))
+        vcodec = format.get("vcodec")
+        acodec = format.get("acodec")
+        if acodec == "none" or (vbr and not abr):
+            pre_index = "vbr"
+            index = int(tbr)
+        elif vcodec == "none" or (abr and not vbr):
+            pre_index = "abr"
+            index = int(tbr)
+        elif tbr:
+            pre_index = "br"
+            index = int(tbr)
+        else:
+            pre_index = ""
+
+        format_id = "%s-%s-%s%d" % (type, transport, pre_index, index)
+        format["format_id"] = format_id
```

Comment:
unnecessary

Reply:
IMO this is necessary if users actually want to know what they are downloading. (see the note about `preference`)

Here is a comparison of formats with and without `_gen_format_id()` (for video `https://tube.tugraz.at/paella/ui/watch.html?id=6b78b5ef-605d-423c-9b6b-5e6260121f26`):

With format ids:
```
presenter-dash-abr129      m4a        audio only  [eng] DASH audio,  129k, m4a_dash container, mp4a.40.2@129k (44100Hz), ~83.55GiB
presenter-dash-abr162      m4a        audio only  [eng] DASH audio,  162k, m4a_dash container, mp4a.40.2@162k (44100Hz), ~104.92GiB
presenter-dash-abr196      m4a        audio only  [eng] DASH audio,  196k, m4a_dash container, mp4a.40.2@196k (44100Hz), ~126.53GiB
presenter-dash-vbr2134     mp4        768x432     DASH video, 2134k, mp4_dash container, avc1.4d401e@2134k, 30fps, video only, ~1.35TiB
presenter-dash-vbr2150     mp4        1280x720    DASH video, 2150k, mp4_dash container, avc1.4d401f@2150k, 30fps, video only, ~1.36TiB
presenter-dash-vbr2267     mp4        1920x1080   DASH video, 2267k, mp4_dash container, avc1.4d4028@2267k, 30fps, video only, ~1.43TiB
presenter-https-br2134     mp4        768x432     2134kvideo@2004k, 30fps, audio@129k, ~1.35TiB
presenter-hls-br2134       mp4        768x432     2134kvideo@2134k, audio@  0k, ~1.35TiB
presenter-https-br2150     mp4        1280x720    2150kvideo@1988k, 30fps, audio@162k, ~1.36TiB
presenter-hls-br2150       mp4        1280x720    2150kvideo@2150k, audio@  0k, ~1.36TiB
presenter-https-br2267     mp4        1920x1080   2267kvideo@2071k, 30fps, audio@196k, ~1.43TiB
presenter-hls-br2267       mp4        1920x1080   2267kvideo@2267k, audio@  0k, ~1.43TiB
presentation-dash-vbr270   mp4        512x320     DASH video,  270k, mp4_dash container, avc1.4d4015@ 270k, 30fps, video only, ~174.28GiB
presentation-dash-vbr481   mp4        852x532     DASH video,  481k, mp4_dash container, avc1.4d401f@ 481k, 30fps, video only, ~310.53GiB
presentation-dash-vbr699   mp4        1280x800    DASH video,  699k, mp4_dash container, avc1.4d4020@ 699k, 30fps, video only, ~451.63GiB
presentation-https-vbr270  mp4        512x320      270k, 30fps
presentation-https-vbr481  mp4        852x532      481k, 30fps
presentation-https-vbr699  mp4        1280x800     699k, 30fps
presentation-hls-br270     mp4        512x320      270kvideo@ 270k, audio@  0k, ~174.28GiB
presentation-hls-br481     mp4        852x532      481kvideo@ 481k, audio@  0k, ~310.53GiB
presentation-hls-br699     mp4        1280x800     699kvideo@ 699k, audio@  0k, ~451.63GiB
```

Without format ids:
```
p0aa0br129446   m4a        audio only  [eng] DASH audio,  129k, m4a_dash container, mp4a.40.2@129k (44100Hz), ~83.55GiB
p0aa0br162563   m4a        audio only  [eng] DASH audio,  162k, m4a_dash container, mp4a.40.2@162k (44100Hz), ~104.92GiB
p0aa0br196045   m4a        audio only  [eng] DASH audio,  196k, m4a_dash container, mp4a.40.2@196k (44100Hz), ~126.53GiB
p0va0br2134088  mp4        768x432     DASH video, 2134k, mp4_dash container, avc1.4d401e@2134k, 30fps, video only, ~1.35TiB
p0va0br2150658  mp4        1280x720    DASH video, 2150k, mp4_dash container, avc1.4d401f@2150k, 30fps, video only, ~1.36TiB
p0va0br2267088  mp4        1920x1080   DASH video, 2267k, mp4_dash container, avc1.4d4028@2267k, 30fps, video only, ~1.43TiB
6               mp4        768x432     2134kvideo@2004k, 30fps, audio@129k, ~1.35TiB
2134            mp4        768x432     2134kvideo@2134k, audio@  0k, ~1.35TiB
8               mp4        1280x720    2150kvideo@1988k, 30fps, audio@162k, ~1.36TiB
2150            mp4        1280x720    2150kvideo@2150k, audio@  0k, ~1.36TiB
10              mp4        1920x1080   2267kvideo@2071k, 30fps, audio@196k, ~1.43TiB
2267            mp4        1920x1080   2267kvideo@2267k, audio@  0k, ~1.43TiB
p0va0br270028   mp4        512x320     DASH video,  270k, mp4_dash container, avc1.4d4015@ 270k, 30fps, video only, ~174.28GiB
p0va0br481131   mp4        852x532     DASH video,  481k, mp4_dash container, avc1.4d401f@ 481k, 30fps, video only, ~310.53GiB
p0va0br699756   mp4        1280x800    DASH video,  699k, mp4_dash container, avc1.4d4020@ 699k, 30fps, video only, ~451.63GiB
15              mp4        512x320      270k, 30fps
16              mp4        852x532      481k, 30fps
17              mp4        1280x800     699k, 30fps
270             mp4        512x320      270kvideo@ 270k, audio@  0k, ~174.28GiB
481             mp4        852x532      481kvideo@ 481k, audio@  0k, ~310.53GiB
699             mp4        1280x800     699kvideo@ 699k, audio@  0k, ~451.63GiB
```

It's almost impossible to discern whether a format is a PRESENTATION or PRESENTER format without labelling them.

Additionally, some transports seem to be slower to download than others (the full video downloads over https usually go far quicker than the chunked HLS or DASH downloads, at least in my experience).

Also, please disregard the sizes listed next to the formats. I know they are completely wrong, I'm currently trying to figure out what causes them to be off.